index.js

```diff
@@ -3,89 +3,115 @@
+    console.log('info in ghost-s3', info);
 };
 
-S3Store.prototype.getTargetDir = function() {
-    var MONTHS = [
-        'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
-        'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'
-    ];
-    var now = new Date();
-    return now.getFullYear() + '/' + MONTHS[now.getMonth()] + '/';
+function getTargetDir() {
+    var now = moment();
+    return now.format('YYYY/MM/');
 };
 
-S3Store.prototype.getTargetName = function(image, targetDir) {
-    var ext = path.extname(image.name),
-        name = path.basename(image.name, ext).replace(/\W/g, '_');
+function getTargetName(image, targetDir) {
+    var ext = path.extname(image.name);
+    var name = path.basename(image.name, ext).replace(/\W/g, '_');
 
     return targetDir + name + '-' + Date.now() + ext;
 };
 
-S3Store.prototype.logError = function(error) {
-    console.log('error in ghost-s3', error);
+function validOptions(opts) {
+    return (opts.accessKeyId && \
+        opts.secretAccessKey && \
+        opts.bucket && \
+        opts.region);
+}
+
+S3Store.prototype.save = function(image) {
+    if (!validOptions(options)) {
+      return Bluebird.reject('ghost-s3 is not configured');
+    }
+
+    var targetDir = getTargetDir();
+    var targetFilename = getTargetName(image, targetDir);
+
+    var s3 = new AWS.S3({
+        accessKeyId: options.accessKeyId,
+        secretAccessKey: options.secretAccessKey,
+        bucket: options.bucket,
+        region: options.region
+    });
+
+    return readFileAsync(image.path)
+        .then(function(buffer) {
+            var params = {
+                ACL: 'public-read',
+                Bucket: options.bucket,
+                Key: targetFilename,
+                Body: buffer,
+                ContentType: image.type,
+                CacheControl: 'max-age=' + (1000 * 365 * 24 * 60 * 60) // 365 days
+            };
+
+            return s3.putObject(params).promise();
+        })
+        .tap(function() {
+            logInfo('ghost-s3', 'Temp uploaded file path: ' + image.path);
+        })
+        .then(function(results) {
+            var awsPath = getAwsPath(options.bucket);
+            return Bluebird.resolve(awsPath + targetFilename);
+        })
+        .catch(function(err) {
+            logError(err);
+            throw err;
+        });
 };
 
-S3Store.prototype.logInfo = function(info) {
-    console.log('info in ghost-s3', info);
+// middleware for serving the files
+S3Store.prototype.serve = function() {
+    var s3 = new AWS.S3({
+        accessKeyId: options.accessKeyId,
+        secretAccessKey: options.secretAccessKey,
+        bucket: options.bucket,
+        region: options.region
+    });
+
+    return function (req, res, next) {
+        var params = {
+            Bucket: options.bucket,
+            Key: req.path.replace(/^\//, '')
+        };
+
+        s3.getObject(params)
+            .on('httpHeaders', function(statusCode, headers, response) {
+                res.set(headers);
```

Comment:
This part was tricky. If you request a file from s3, you get a bunch of headers that are saved there, like `CacheControl`. This makes it so you send those back out if you're proxying through ghost.